File: rb5_ros2_vision/src/undistort_example.py
import sys
import os
from os import path as osp
import cv2
import yaml
import numpy as np
from timeit import default_timer as timer


def read_image_from_dir(img_dir):
    img_list = []

    files = os.listdir(img_dir)

    for file in files:
        if file.endswith('.png') or file.endswith('.jpg'):
            img_path = osp.join(img_dir, file)
            img = cv2.imread(img_path)
            img_list.append(img)

    return img_list


def parse_calibration_yaml(yaml_path):
    calibration_dict = {}
    with open(yaml_path) as yaml_fp:
        raw_data = yaml.safe_load(yaml_fp)
        
        calibration_dict["width"] = raw_data["image_width"]
        calibration_dict["height"] = raw_data["image_height"]
        calibration_dict["camera_matrix"]=np.array(
            raw_data["camera_matrix"]["data"]
        ).reshape(
            raw_data["camera_matrix"]["cols"], 
            raw_data["camera_matrix"]["rows"])
        calibration_dict["distortion_coefficients"] = np.array(
            raw_data["distortion_coefficients"]["data"]
        )

        return calibration_dict


def undistort_slow(img_list, mtx, dist, w, h, newcameramtx, roi):
    img_list_undist = []
    x, y, w, h = roi
    
    for img in img_list:
        img_undist = cv2.undistort(img, mtx, dist, None, newcameramtx)
        
        # crop the image    
        img_undist = img_undist[y:y+h, x:x+w]
        
        img_list_undist.append(img_undist)
    

    
    return img_list_undist


def undistort_fast(img_list, mtx, dist, w, h, newcameramtx, roi):
    img_list_undist = []
    x, y, w, h = roi
    mapx, mapy = cv2.initUndistortRectifyMap(mtx, dist, None, newcameramtx, (w,h), 5)

    for img in img_list:
        img_undist = cv2.remap(img, mapx, mapy, cv2.INTER_LINEAR)
        
        # crop the image    
        img_undist = img_undist[y:y+h, x:x+w]
        
        img_list_undist.append(img_undist)
    
    return img_list_undist


def display_image_list(img_list):
    for img in img_list:
        cv2.imshow('image', img)
        cv2.waitKey(0)

def main():
    print(cv2.getBuildInformation())
    img_dir = "/home/henry/Documents/data/000-calibration"
    yaml_path = osp.join(img_dir, "ost.yaml")
    img_list = read_image_from_dir(img_dir)
    calibration_dict = parse_calibration_yaml(yaml_path)
    
    mtx = calibration_dict["camera_matrix"]
    dist = calibration_dict["distortion_coefficients"]
    w = calibration_dict["width"]
    h = calibration_dict["height"]

    newcameramtx, roi = cv2.getOptimalNewCameraMatrix(mtx, dist, (w,h), 1, (w,h))

    start = timer()
    img_list_undist_slow = undistort_slow(img_list, mtx, dist, w, h, newcameramtx, roi)
    print("slow: ", timer() - start)

    start = timer()
    img_list_undist_fast = undistort_fast(img_list, mtx, dist, w, h, newcameramtx, roi)
    print("fast: ", timer() - start)

    display_image_list(img_list_undist_fast)


main()

File: rb5_ros2_vision/src/undistort_example.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <ctime>
#include <cstdio>
#include <sys/time.h>

#include <opencv2/core.hpp>
#include <opencv2/core/utility.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/calib3d.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/videoio.hpp>
#include <opencv2/highgui.hpp>

using namespace cv;
using namespace std;

class Settings
{
public:
    Settings() : goodInput(false) {}

    void write(FileStorage& fs) const                        //Write serialization for this class
    {
        fs << "image_width"  << image_width;
        fs << "image_height" << image_height;
        fs << "camera_matrix" << camera_matrix;
        fs << "distortion_coefficients" << distortion_coefficients;
    }

    void read(const FileNode& node)                          //Read serialization for this class
    {
        node["image_width" ] >> image_width;
        node["image_height"] >> image_height;
        node["camera_matrix"] >> camera_matrix;
        node["distortion_coefficients"] >> distortion_coefficients;

        validate();
    }

    void read(const FileStorage& fs)                          //Read serialization for this class
    {
        fs["image_width" ] >> image_width;
        fs["image_height"] >> image_height;
        fs["camera_matrix"] >> camera_matrix;
        fs["distortion_coefficients"] >> distortion_coefficients;

        validate();
    }

    void validate()
    {
        goodInput = true;
        if (image_width <= 0 || image_height <= 0)
        {
            cerr << "Invalid image size: " << image_width << " " << image_height << endl;
            goodInput = false;
        }
    }

public:
    int image_width;
    int image_height;
    Mat camera_matrix;
    Mat distortion_coefficients;
    Mat rectification_matrix;
    Mat projection_matrix;
    bool goodInput;

    bool useFisheye;             // use fisheye camera model for calibration

};

int main(int argc, char* argv[])
{

    const String keys
        = "{help h usage ? |           | print this message            }"
          "{@settings      |default.xml| input setting file            }";
    CommandLineParser parser(argc, argv, keys);
    parser.about("This is a camera calibration sample.\n"
                 "Usage: camera_calibration [configuration_file -- default ./default.xml]\n"
                 "Near the sample file you'll find the configuration file, which has detailed help of "
                 "how to edit it. It may be any OpenCV supported file format XML/YAML.");
    if (!parser.check()) {
        parser.printErrors();
        return 0;
    }

    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }

    std::vector<cv::String> imageList;
    std::string imageDir("/home/henry/Documents/data/000-calibration/");
    std::string imageSuffix("*.png");
    glob((imageDir+imageSuffix).c_str(), imageList, false);

    //! [file_read]
    Settings s;
    const string inputSettingsFile = parser.get<string>(0);
    std::cout << "input file name: " << inputSettingsFile << std::endl;
    FileStorage fs(inputSettingsFile, FileStorage::READ); // Read the settings
    if (!fs.isOpened())
    {
        cout << "Could not open the configuration file: \"" << inputSettingsFile << "\"" << endl;
        parser.printMessage();
        return -1;
    }
    int width;
    s.read(fs);
    fs.release();                                         // close Settings file
    //! [file_read]

    //FileStorage fout("settings.yml", FileStorage::WRITE); // write config as YAML
    //fout << "Settings" << s;

    if (!s.goodInput)
    {
        cout << "Invalid input detected. Application stopping. " << endl;
        return -1;
    }



    // if( mode == CALIBRATED && s.showUndistorted )
    // {
    //     Mat temp = view.clone();
    //     if (s.useFisheye)
    //     {
    //         Mat newCamMat;
    //         fisheye::estimateNewCameraMatrixForUndistortRectify(cameraMatrix, distCoeffs, imageSize,
    //                                                             Matx33d::eye(), newCamMat, 1);
    //         cv::fisheye::undistortImage(temp, view, cameraMatrix, distCoeffs, newCamMat);
    //     }
    //     else
    //         undistort(temp, view, cameraMatrix, distCoeffs);
    // }
    // -----------------------Show the undistorted image for the image list ------------------------
    //! [show_results]

    Mat cameraMatrix = s.camera_matrix.clone();
    Mat distCoeffs = s.distortion_coefficients.clone();
    Size imageSize;
    imageSize.width = s.image_width;
    imageSize.height = s.image_height;
    const char ESC_KEY = 27;

    if( !cameraMatrix.empty())
    {
        Mat view, rview, map1, map2;

        if (s.useFisheye)
        {
            Mat newCamMat;
            fisheye::estimateNewCameraMatrixForUndistortRectify(cameraMatrix, distCoeffs, imageSize,
                                                                Matx33d::eye(), newCamMat, 1);
            fisheye::initUndistortRectifyMap(cameraMatrix, distCoeffs, Matx33d::eye(), newCamMat, imageSize,
                                             CV_16SC2, map1, map2);
        }
        else
        {
            initUndistortRectifyMap(
                cameraMatrix, distCoeffs, Mat(),
                getOptimalNewCameraMatrix(cameraMatrix, distCoeffs, imageSize, 0, imageSize, 0), imageSize,
                CV_16SC2, map1, map2);
        }

        std::cout << "imageSize: " << imageSize << std::endl;



        std::vector<Mat> imageReadList;

        for(size_t i = 0; i < imageList.size(); i++ )
        {
            view = imread(imageList[i], IMREAD_COLOR);
            if(view.empty())
                continue;
            imageReadList.push_back(view);
        }
        
        timeval begin, end;
        long seconds, u_seconds;
        gettimeofday(&begin, 0);

        for(size_t i = 0; i < imageReadList.size(); i++ )
        {
            remap(imageReadList[i], rview, map1, map2, INTER_LINEAR);
            // undistort(imageReadList[i], rview, cameraMatrix, distCoeffs);
            namedWindow("Image", WINDOW_GUI_NORMAL);
            imshow("Image", rview);
            char c = (char)waitKey();
            if( c  == ESC_KEY || c == 'q' || c == 'Q' )
                break;

        }

        gettimeofday(&end, 0);
        seconds = end.tv_sec - begin.tv_sec;
        u_seconds = end.tv_usec - begin.tv_usec;
        std::cout << seconds << "s " << u_seconds << "us" << std::endl;
    }
    //! [show_results]

    return 0;
}